packages/flutter/lib/src/services/asset_manifest.dart

```diff
@@ -0,0 +1,94 @@
+  /// [AssetManifest] object from that data.
+  static Future<AssetManifest> loadFromAssetBundle(AssetBundle bundle) {
+    return bundle.loadStructuredBinaryData(_kAssetManifestFilename, _AssetManifestBin.fromStandardMessageCodecMessage);
+  }
+
+  /// Loads asset manifest data from the root bundle and creates an
+  /// [AssetManifest] from that data.
+  static Future<AssetManifest> loadFromRootBundle() {
+    return loadFromAssetBundle(rootBundle);
+  }
+
+  /// Lists the keys of all known assets, not including asset variants.
+  Iterable<String> listAssets() {
+    throw UnimplementedError();
+  }
+
+  /// Gets available variants of an asset.
+  Iterable<AssetVariant> getAssetVariants(String key) {
+    throw UnimplementedError();
+  }
+}
+
+/// Parses the binary asset manifest into a data structure that's easier to work with.
+///
+/// The asset manifest is a map of asset files to a list of objects containing
+/// information about variants of that asset.
+///
+/// The entries with each variant object are:
+///  - "asset": the location of this variant to load it from.
+///  - "dpr": The device-pixel-ratio that the asset is best-suited for.
+///
+/// New fields could be added to this object schema to support new asset variation
+/// features, such as themes, locale/region support, reading directions, and so on.
+class _AssetManifestBin implements AssetManifest {
+  _AssetManifestBin(Map<Object?, Object?> standardMessageData): _data = standardMessageData;
+
+  factory _AssetManifestBin.fromStandardMessageCodecMessage(ByteData message) {
+    final dynamic data = const StandardMessageCodec().decodeMessage(message);
+    return _AssetManifestBin(data as Map<Object?, Object?>);
+  }
+
+  final Map<Object?, Object?> _data;
+  final Map<String, Iterable<AssetVariant>> _typeCastedData = <String, Iterable<AssetVariant>>{};
+
+  @override
+  Iterable<AssetVariant> getAssetVariants(String key) {
+    // We lazily delay typecasting to prevent a performance hiccup when parsing
+    // large asset manifests.
+    if (!_typeCastedData.containsKey(key)) {
+      _typeCastedData[key] = ((_data[key] ?? <Object?>[]) as List<Object?>)
+        .cast<Map<Object?, Object?>>()
+        .map((Map<Object?, Object?> data) => AssetVariant(
+            key: data['asset']! as String,
+            targetDevicePixelRatio: data['dpr']! as double,
+        ));
+
+      _data.remove(key);
+    }
+    return _typeCastedData[key]!;
+  }
+
+  @override
+  Iterable<String> listAssets() {
+    return <String>[..._data.keys.cast<String>(), ..._typeCastedData.keys];
+  }
+}
+
+/// Contains information about an asset that is a variant of another asset.
+class AssetVariant {
+  /// Creates an object containing information about an asset variant.
+  AssetVariant({
+    required this.key,
+    required this.targetDevicePixelRatio,
+  });
+
+  /// The device pixel ratio that the asset is most ideal for, if any.
+  final double? targetDevicePixelRatio;
```

Comment:
This would be null for any asset not in a 1x/2x subfolder right?

I would consider linking to the website assets page too.

Reply:
> This would be null for any asset not in a 1x/2x subfolder right?

Yup, exactly. I chose to be passive and not doc comment that specifically--just in case things change in the future. A link to the doc site seems like a good idea, though.

Reply:
Not documenting how an API behaves is not a reasonable defense for changing the API in the future. If we need to change the behavior, we have criteria on the wiki for determining whether or not it qualifies as a breaking change.

Please document the behavior

Reply:
If by "API" you are referring to how the ratio is derived from the file path, then I see your point. I made the doc string more informative and added a docsite link.